### ISSUE TITLE
Compiled tensor subclass can fail when graph break happens in the wrong place

### ISSUE THREAD

**Issue by soulitzer** [pytorch/pytorch#133644]
### 🐛 Describe the bug

```python
import torch

def fn(nt):
    tmp1 = nt.clone()
    tmp2 = nt.clone()
    torch._dynamo.graph_break()
    return tmp1.clone(), tmp2.clone().values().detach()

compiled_fn =torch.compile(fn, backend="aot_eager")

values = torch.rand(6, 8, requires_grad=True, device="cpu", dtype=torch.float32)
offsets = torch.tensor([0, 2, 3, 6], device="cpu")
nt = torch.nested.nested_tensor_from_jagged(
  values, offsets, lengths=None, min_seqlen=1, max_seqlen=8
)
out, _ = compiled_fn(nt)
out.values().sum().backward()
```

When custom autograd Function (1) returns more than one output, (2) during backward, we don't invoke backward wrt all the outputs (unless ctx.set_materialize_grads=False) custom autograd Function materializes zero gradients in the shape of the outputs. This is problematic for NT today because NT does not support factory functions. 

In the repro above, shows that this problem can show up even if the user does not explicitly not-backward-through-all-outputs. Depending on where a graph break occured, it is possible to accidentally create a CompiledFunction returning more than one NT, such that when gradients flow back, they are None wrt one of the outputs. 

Solutions:
- Support NT factory functions
- others?

### Versions

main

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @cpuhrsch @jbschlosser @bhosmer @drisspg @davidberard98 @YuqingJ @chauhang @penguinwu

**Comment by soulitzer** [comment]
This is also a general subclass issue as most (all?) subclasses don't have factory function support. Unlike NJT, for two_tensor, custom autograd Function cannot tell from the sizes alone that it needs to create a zeros TwoTensor. 

```python
import torch
from torch.testing._internal.two_tensor import TwoTensor

class Split(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        return x.a, x.b

    @staticmethod
    def backward(ctx, ga, gb):
        return TwoTensor(ga.clone(), gb.clone())

def fn(nt):
    tmp1 = nt.clone()
    tmp2 = nt.clone()
    torch._dynamo.graph_break()
    return tmp1.clone(), Split.apply(tmp2.clone())[0].detach()

compiled_fn =torch.compile(fn, backend="aot_eager")

a = torch.tensor([1., 2.], requires_grad=True)
b = torch.tensor([1., 2.], requires_grad=True)
x = TwoTensor(a, b).requires_grad_(True)
out, _ = compiled_fn(x)
Split.apply(out)[0].sum().backward()
```

```
AssertionError:     We incorrectly attempted to compile the backward with incorrect subclass metadata.
    If you run into this error, please file an issue.
    Expected grad_output types: [<class 'torch.testing._internal.two_tensor.TwoTensor'>, <class 'torch.testing._internal.two_tensor.TwoTensor'>]
    Got grad_output types: [<class 'torch.testing._internal.two_tensor.TwoTensor'>, <class 'torch.Tensor'>]
```

**Comment by bdhirsh** [comment]
The current PoR is to fix the "incorrect subclass metadata" error as part of https://github.com/pytorch/pytorch/issues/91469:

(1) we will trace the "backward section" of the joint graph into a pre-dispatch graph, that has not yet burned in decisions about whether the tangent inputs are subclasses

(2) at runtime of the backward, we will take the backward, pre-dispatch graph and re-trace it into a proper post-dispatch ATen graph with our correct tangent inputs.

**Comment by bdhirsh** [comment]
From offline discussion with @soulitzer: one difference between the case with NJT's vs other subclasses here is that:

(1) For subclasses in general, autograd requires that fw outputs and their gradients have the **same shape** but not necessarily the same subclass type. This is what makes it "okay" for autograd.Function to construct plain tensors of zeros as tangents, when the corresponding forward outputs are tensor subclasses (as long as it gives them the same shape)

(2) for NJT though, its shape includes a `nested_int`. We cannot construct a plain tensor that holds a "nested int" size today. So in order to satisfy autograd and construct a tangent with the same shape as the forward output, we need autograd's call to `torch.zeros(nested_int)` to properly construct an NJT.

That seems like it will require doing the work to have `nested_int` reference the offsets tensor directly, so factory functions can use the offsets to faithfully construct an NJT of the right size.

**Comment by bdhirsh** [comment]
@soulitzer is this still hi-pri?

**Comment by soulitzer** [comment]
For NJT this will be addressed soon via a short-term fix @jbschlosser, happy with removing high-pri with that in

**Comment by jbschlosser** [comment]
Short-term fix landed in #136875; eventually we should switch to using factory functions when NJT supports them.
